sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1407,7 +1407,15 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def hint(name: String, parameters: Any*): Dataset[T] = withTypedPlan {
-    UnresolvedHint(name, parameters, logicalPlan)
+    // parse string parameters into Expressions as ResolveHint requires all the parameters to be
+    // expressions except the first one could be numeric. This logic matches how sql hint is parsed
+    // and makes caller easier to pass string parameters in hint specification, especially for
+    // other language bindings, such as PySpark.
+    val pars = parameters.map {
+      case s: String => sparkSession.sessionState.sqlParser.parseExpression(s)
```

Comment:
I don't quite get this. Why do we only handle strings? If all params need to be expressions, can't we just wrap it with `Literal`?

Reply:
This is too much breaking change. Previously the strings were literals and now they are treated as columns.

Reply:
cc @ulysses-you mind taking a look please?

Reply:
I'm fine with this idea. However, I think we should treat all input parameters as expression like what AstBuilder does rathen than only handle string type. Besides, we should add one more config, since the hint framework is a kind of extension, developers may have their own resolution rules.

Reply:
I mean we can add something like this, and by default we can tune this config off.

```scala
val parsed = if (parseHintParameterAsExpression) {
  parseExpression ...
} else {
  origin
}
UnresolvedHint(name, parsed, child)
```